exec/java-exec/src/main/java/org/apache/drill/exec/expr/stat/ParquetIsPredicate.java

```diff
@@ -62,70 +62,72 @@ private ParquetIsPredicate(LogicalExpression expr, BiPredicate<Statistics<C>, Ra
-
-      private boolean isArray(LogicalExpression expression) {
-        if (expression instanceof TypedFieldExpr) {
-          TypedFieldExpr typedFieldExpr = (TypedFieldExpr) expression;
-          SchemaPath schemaPath = typedFieldExpr.getPath();
-          return schemaPath.isArray();
-        }
-        return false;
-      }
-
-      @Override
-      public boolean canDrop(RangeExprEvaluator<C> evaluator) {
+      (exprStat, evaluator) -> {
         // for arrays we are not able to define exact number of nulls
         // [1,2,3] vs [1,2] -> in second case 3 is absent and thus it's null but statistics shows no nulls
-        return !isArray && super.canDrop(evaluator);
-      }
-    };
+        if(expr instanceof TypedFieldExpr) {
+          TypedFieldExpr typedFieldExpr = (TypedFieldExpr) expr;
+          if (typedFieldExpr.getPath().isArray()) {
+            return RowsMatch.SOME;
+          }
+        }
+        if (hasNoNulls(exprStat)) {
+          return RowsMatch.NONE;
+        }
+        return isAllNulls(exprStat, evaluator.getRowCount()) ? RowsMatch.ALL : RowsMatch.SOME;
+      });
   }
 
   /**
    * IS NOT NULL predicate.
    */
   private static <C extends Comparable<C>> LogicalExpression createIsNotNullPredicate(LogicalExpression expr) {
     return new ParquetIsPredicate<C>(expr,
-        //if there are all nulls  -> canDrop
-        (exprStat, evaluator) -> isAllNulls(exprStat, evaluator.getRowCount())
+      (exprStat, evaluator) -> isAllNulls(exprStat, evaluator.getRowCount()) ? RowsMatch.NONE : checkNull(exprStat)
     );
   }
 
   /**
    * IS TRUE predicate.
    */
   private static LogicalExpression createIsTruePredicate(LogicalExpression expr) {
-    return new ParquetIsPredicate<Boolean>(expr, (exprStat, evaluator) ->
-        //if max value is not true or if there are all nulls  -> canDrop
-        isAllNulls(exprStat, evaluator.getRowCount()) || exprStat.hasNonNullValue() && !((BooleanStatistics) exprStat).getMax()
-    );
+    return new ParquetIsPredicate<Boolean>(expr, (exprStat, evaluator) -> {
+      if (isAllNulls(exprStat, evaluator.getRowCount()) || exprStat.hasNonNullValue() && !((BooleanStatistics)exprStat).getMax()) {
+        return RowsMatch.NONE;
+      }
+      return ((BooleanStatistics)exprStat).getMin() && ((BooleanStatistics)exprStat).getMax() ? checkNull(exprStat) : RowsMatch.SOME;
+    });
   }
 
   /**
    * IS FALSE predicate.
    */
   private static LogicalExpression createIsFalsePredicate(LogicalExpression expr) {
     return new ParquetIsPredicate<Boolean>(expr, (exprStat, evaluator) ->
-        //if min value is not false or if there are all nulls  -> canDrop
-        isAllNulls(exprStat, evaluator.getRowCount()) || exprStat.hasNonNullValue() && ((BooleanStatistics) exprStat).getMin()
+      exprStat.hasNonNullValue() && ((BooleanStatistics)exprStat).getMin() || isAllNulls(exprStat, evaluator.getRowCount()) ? RowsMatch.NONE : checkNull(exprStat)
```

Comment:
Is it necessary to check for max to be False before returning ALL? The same for "is not true/false".

Reply:
Yes. Necessary for test file blnTbl/0_0_2.parquet where we have ST:[min: false, max: true, num_nulls: 1]. In this case, the file is discarded, which is wrong since it contains 1 false.

Reply:
If yes, why is it missing? The use case provided does not apply. I refer to a case where `is false` should return ALL: min is false, max is false, number of nulls is 0.

Reply:
Exact. Corrected.